pkg/provisioner/dependency.go

```diff
@@ -191,3 +86,83 @@ func kind(tpl *bytes.Buffer, env v1alpha1.Environment) error {
 	}
 	return kind.Execute(tpl, env)
 }
+
+// DependencySolver is a struct that holds the dependency graph
+type DependencySolver struct {
+	Dependencies []ProvisionFunc
+	env          v1alpha1.Environment
+}
+
+func NewDependencySolver(env v1alpha1.Environment) *DependencySolver {
+	return &DependencySolver{
+		env: env,
+	}
+}
+
+func (d *DependencySolver) withKubernetes() {
+	switch d.env.Spec.Kubernetes.KubernetesInstaller {
+	case "kubeadm":
+		d.Dependencies = append(d.Dependencies, functions["kubeadm"])
+	case "kind":
+		d.Dependencies = append(d.Dependencies, functions["kind"])
+	case "microk8s":
+		// reset the graph to only include microk8s
+		d.Dependencies = nil
+		d.Dependencies = append(d.Dependencies, functions["microk8s"])
+	default:
+		// default to kubeadm if KubernetesInstaller is empty
```

Comment:
Question: Why is the default not handled elsewhere? Should `d.env.Spec.Kubernetes.KubernetesInstaller` not be set to a default before calling this function?

Reply:
Until this point we have not modified `d.env` , and after this step is not needed.  I think is ok to have it here. unless there's a reason to set the defaults at an earlier stage